Can't pass dot/hidden files on command line

Hello,
since JavaScript config files are supported now, can you please consider adding a (at least optional) way to ask `eslint` to lint also hidden/dot files (e.g. `.eslintrc.js` or `.eslintrc.json` with the JSON plugin) if a name to a directory is passed (e.g. `eslint .`)?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29459770-optionally-lint-dot-hidden-files?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

@nzakas, sure, here it is:

```sh
eslint --debug ./eslintrc.js
```

```
  eslint:cli Running on files +0ms
  eslint:ignored-paths Looking for ignore file in /home/dp/Projects/sdk/cjs/eslint-config +28ms
  eslint:ignored-paths Could not find ignore file in cwd +1ms
  eslint:cli-engine Processing /home/dp/Projects/sdk/cjs/eslint-config/eslintrc.js +4ms
  eslint:cli-engine Linting /home/dp/Projects/sdk/cjs/eslint-config/eslintrc.js +1ms
  eslint:config Constructing config for /home/dp/Projects/sdk/cjs/eslint-config/eslintrc.js +1ms
  eslint:config Using .eslintrc and package.json files +0ms
  eslint:config Loading /home/dp/Projects/sdk/cjs/eslint-config/.eslintrc.js +2ms
  eslint:config-file Loading JS config file: /home/dp/Projects/sdk/cjs/eslint-config/.eslintrc.js +1ms
  eslint:config Using /home/dp/Projects/sdk/cjs/eslint-config/.eslintrc.js +187ms
  eslint:config Loading /home/dp/Projects/sdk/package.json +0ms
  eslint:config-file Loading package.json config file: /home/dp/Projects/sdk/package.json +1ms
  eslint:config-file Loading JSON config file: /home/dp/Projects/sdk/package.json +0ms
  eslint:config Merging command line environment settings +1ms
  eslint:config-ops Apply environment settings to config +0ms
  eslint:config-ops Creating config for environment commonjs +0ms
  eslint:cli-engine Linting complete in: 386ms +189ms

/home/dp/Projects/sdk/cjs/eslint-config/eslintrc.js
  19:9   warning  Extra space after key 'comma-dangle'               key-spacing
  19:25  warning  Missing space before value for key 'comma-dangle'  key-spacing
  19:26  error    Strings must use singlequote                       quotes

✖ 3 problems (1 error, 2 warnings)
```

--- a/lib/util/glob-util.js
+++ b/lib/util/glob-util.js
@@ -104,18 +104,11 @@ function resolveFileGlobPatterns(patterns, options) {
  * @returns {string[]} Resolved absolute filenames.
  */
 function listFilesToProcess(globPatterns, options) {
+    options = options || { ignore: true };
     const files = [],
         added = {};
 
     const cwd = (options && options.cwd) || process.cwd();
-
-    options = options || { ignore: true, dotfiles: true };
-    const ignoredPaths = new IgnoredPaths(options);
-    const globOptions = {
-        nodir: true,
-        cwd
-    };
-    const shouldIgnore = ignoredPaths.getIgnoredFoldersGlobChecker();
 
     /**
      * Executes the linter on a file defined by the `filename`. Skips
@@ -123,9 +116,10 @@ function listFilesToProcess(globPatterns, options) {
      * @param {string} filename The file to be processed
      * @param {boolean} shouldWarnIgnored Whether or not a report should be made if
      *                                    the file is ignored
+     * @param {IgnoredPaths} ignoredPaths An instance of IgnoredPaths
      * @returns {void}
      */
-    function addFile(filename, shouldWarnIgnored) {
+    function addFile(filename, shouldWarnIgnored, ignoredPaths) {
         let ignored = false;
         let isSilentlyIgnored;
 
@@ -160,10 +154,24 @@ function listFilesToProcess(globPatterns, options) {
         const file = path.resolve(cwd, pattern);
 
         if (shell.test("-f", file)) {
-            addFile(fs.realpathSync(file), !shell.test("-d", file));
+            const ignoredPaths = new IgnoredPaths(options);
+
+            addFile(fs.realpathSync(file), !shell.test("-d", file), ignoredPaths);
         } else {
+
+            // regex to find .hidden or /.hidden patterns, but not ./relative or ../relative
+            const globIncludesDotfiles = /(?:(?:^\.)|(?:[\/\\]\.))[^\/\\\.].*/.test(pattern);
+
+            const ignoredPaths = new IgnoredPaths(Object.assign({}, options, {dotfiles: options.dotfiles || globIncludesDotfiles}));
+            const shouldIgnore = ignoredPaths.getIgnoredFoldersGlobChecker();
+            const globOptions = {
+                nodir: true,
+                dot: true,
+                cwd,
+            };
+
             new GlobSync(pattern, globOptions, shouldIgnore).found.forEach(function(globMatch) {
-                addFile(path.resolve(cwd, globMatch), false);
+                addFile(path.resolve(cwd, globMatch), false, ignoredPaths);
             });
         }
     });

--- a/tests/fixtures/cli-engine/hidden/.hiddenfolder/double-quotes.js
+++ b/tests/fixtures/cli-engine/hidden/.hiddenfolder/double-quotes.js
@@ -1,0 +1,2 @@
+var bar = "<div>Hello world!</div>";
+bar(bar);

--- a/tests/lib/cli-engine.js
+++ b/tests/lib/cli-engine.js
@@ -523,6 +523,22 @@ describe("CLIEngine", function() {
             assert.equal(report.results[0].messages[0].message, expectedMsg);
         });
 
+        it("should report on globs with explicit inclusion of dotfiles, even though ignored by default", function() {
+
+            engine = new CLIEngine({
+                cwd: getFixturePath("cli-engine"),
+                rules: {
+                    quotes: [2, "single"]
+                }
+            });
+
+            const report = engine.executeOnFiles(["hidden/.hiddenfolder/*.js"]);
+
+            assert.equal(report.results.length, 1);
+            assert.equal(report.results[0].errorCount, 1);
+            assert.equal(report.results[0].warningCount, 0);
+        });
+
         it("should not check default ignored files without --no-ignore flag", function() {
 
             engine = new CLIEngine({
@@ -578,6 +594,26 @@ describe("CLIEngine", function() {
             });
 
             const report = engine.executeOnFiles(["fixtures/files/.bar.js"]);
+
+            assert.equal(report.results.length, 1);
+            assert.equal(report.results[0].warningCount, 0);
+            assert.equal(report.results[0].errorCount, 1);
+            assert.equal(report.results[0].messages[0].ruleId, "quotes");
+        });
+
+        it("should check .hidden files if they are unignored with an --ignore-pattern", function() {
+
+            engine = new CLIEngine({
+                cwd: getFixturePath("cli-engine"),
+                ignore: true,
+                useEslintrc: false,
+                ignorePattern: "!.hidden*",
+                rules: {
+                    quotes: [2, "single"]
+                }
+            });
+
+            const report = engine.executeOnFiles(["hidden/"]);
 
             assert.equal(report.results.length, 1);
             assert.equal(report.results[0].warningCount, 0);

--- a/tests/lib/util/glob-util.js
+++ b/tests/lib/util/glob-util.js
@@ -199,7 +199,9 @@ describe("globUtil", function() {
 
         it("should not return hidden files for standard glob patterns", function() {
             const patterns = [getFixturePath("glob-util", "hidden", "**/*.js")];
-            const result = globUtil.listFilesToProcess(patterns);
+            const result = globUtil.listFilesToProcess(patterns, {
+                cwd: getFixturePath()
+            });
 
             assert.equal(result.length, 0);
         });
@@ -207,8 +209,7 @@ describe("globUtil", function() {
         it("should return hidden files if included in glob pattern", function() {
             const patterns = [getFixturePath("glob-util", "hidden", "**/.*.js")];
             const result = globUtil.listFilesToProcess(patterns, {
-                cwd: getFixturePath(),
-                dotfiles: true
+                cwd: getFixturePath()
             });
 
             const file1 = getFixturePath("glob-util", "hidden", ".foo.js");

